python/ray/autoscaler/cli_logger.py

```diff
@@ -0,0 +1,306 @@
+        tags_str = ""
+        if _tags is not None:
+            tags_list = []
+            for k, v in _tags.items():
+                if v is True:
+                    tags_list += [k]
+                    continue
+                if v is False:
+                    continue
+
+                tags_list += [k + "=" + v]
+            if tags_list:
+                tags_str = cf.reset(
+                    cf.gray(" [{}]".format(", ".join(tags_list))))
+
+        numbering_str = ""
+        if _numbered is not None:
+            chars, i, n = _numbered
+
+            i = str(i)
+            n = str(n)
+
+            numbering_str = cf.gray(chars[0] + i + "/" + n + chars[1]) + " "
+
+        if _no_format:
+            # todo: throw if given args/kwargs?
+            return numbering_str + msg + tags_str
+        return numbering_str + msg.format(*args, **kwargs) + tags_str
+
+    if kwargs:
+        raise ValueError("We do not support printing kwargs yet.")
+
+    res = [msg, *args]
+    res = [str(x) for x in res]
+    return ", ".join(res)
+
+
+class _CliLogger():
+    def __init__(self):
+        self.strip = False
+        self.old_style = True
+        self.color_mode = "auto"
+        self.indent_level = 0
+        self.verbosity = 0
+
+        self.dump_command_output = False
+
+        self.info = {}
+
+    def detect_colors(self):
+        if self.color_mode == "true":
+            self.strip = False
+            return
+        if self.color_mode == "false":
+            self.strip = True
+            return
+        if self.color_mode == "auto":
+            self.strip = sys.stdout.isatty()
+            return
+
+        raise ValueError("Invalid log color setting: " + self.color_mode)
+
+    def newline(self):
+        self._print("")
+
+    def _print(self, msg, linefeed=True):
+        if self.old_style:
+            return
+
+        if self.strip:
+            msg = _strip_codes(msg)
+
+        rendered_message = "  " * self.indent_level + msg
+
+        if not linefeed:
+            sys.stdout.write(rendered_message)
+            sys.stdout.flush()
+            return
+
+        print(rendered_message)
+
+    def indented(self, cls=False):
+        cli_logger = self
+
+        class IndentedContextManager():
+            def __enter__(self):
+                cli_logger.indent_level += 1
+
+            def __exit__(self, type, value, tb):
+                cli_logger.indent_level -= 1
+
+        if cls:
+            # fixme: this does not work :()
+            return IndentedContextManager
+        return IndentedContextManager()
+
+    def timed(self, msg, *args, **kwargs):
+        self.print(msg, *args, **kwargs)
+
+        # todo: implement timer
+        cli_logger = self
+
+        class TimedContextManager():
+            def __enter__(self):
+                cli_logger.indent_level += 1
+
+            def __exit__(self, type, value, tb):
+                cli_logger.indent_level -= 1
+
```

Comment:
these two seem to duplicate; can you just reuse? 

Also since "cls" doesn't work, you could just define it outside right?

Reply:
yeah you are right, this stuff is just artifacts of the past

originally, I tried to make TimedContext a subclass of IndentedContext to remove the code duplication here, using the `cls` thing

really we should just cut TimedContext cause I don't think it will ever get implemented (or just replace it with a call to `self.group(...)`)